### PR TITLE
Animated webp encoder

### LINKED ISSUE #2521: [webp]animated webp image became static after processing
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have verified that I am running the latest version of ImageSharp
- [X] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [X] I have searched [open](https://github.com/SixLabors/ImageSharp/issues) and [closed](https://github.com/SixLabors/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### ImageSharp version

3.0.1

### Other ImageSharp packages and versions

SixLabors.ImageSharp.Drawing(1.0.0)

### Environment (Operating system, version and so on)

Win10 22H2

### .NET Framework version

.NET 6

### Description

[mathiasbynens.be/demo/animated-webp-supported.webp](https://mathiasbynens.be/demo/animated-webp-supported.webp) cannot load animated webp images or save one correctly. Maybe the method I used is wrong.
Really appreciate for your work~ 

### Steps to Reproduce

1.download the example image
2.load with code
 ```c#
            using (var imgStream = new MemoryStream())
            {
                await Request.Body.CopyToAsync(imgStream);
                if (imgStream == null || imgStream.Length == 0)
                {
                    return Fail("image can not be null");
                }
               var  buffer = new byte[imgStream.Length];
                imgStream.Position = 0;
                await imgStream.ReadAsync(buffer, 0, buffer.Length);
                //using(Image img=Image.Load(file))  //At first i wrote this line which turned out img.Frames.Count=1
                using (Image img = Image.Load(new DecoderOptions { MaxFrames = 100 }, file))//this way turned out img.Frames.Count=12
                {
                            var filePath= AppContext.BaseDirectory + Guid.NewGuid().ToString() + ".webp";
                            await img.SaveAsync(filePath);
                }
            }
 ```

### Images

[68747470733a2f2f6d61746869617362796e656e732e62652f64656d6f2f616e696d617465642d776562702d737570706f727465642e77656270.zip](https://github.com/SixLabors/ImageSharp/files/12461797/68747470733a2f2f6d61746869617362796e656e732e62652f64656d6f2f616e696d617465642d776562702d737570706f727465642e77656270.zip)

### PR DESCRIPTION
**Pull request by Poker-sang**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Animated webp encoder
closed #2521

### CODE REVIEW

**Comment by Poker-sang** [comment]
How to start Github Action?

**Comment by JimBobSquarePants** [comment]
Crikey you’re fast!!

Workflows need permission to run from the repository owner. I’ll run it now.

**Comment by Poker-sang** [comment]
I tried to play the encoded out image in my local image viewer and it plays fine. But I'm not sure how to write a good unit test to verify it.

**Comment by JimBobSquarePants** [comment]
I would look at our Gif unit tests. We test using individual frames plus do metadata tests to ensure consistency.

**Comment by JimBobSquarePants** [comment]
If it's not used yeah.

File: src/ImageSharp/Formats/Webp/WebpAnimationEncoder.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+// Licensed under the Six Labors Split License.
+
+namespace SixLabors.ImageSharp.Formats.Webp;
+
+/// <summary>
+/// Encoder for animated webp images.
+/// </summary>
+public class WebpAnimationEncoder
+{
+    // 可能不需要这个屌东西
```
Comment (gfoidl) [inline src/ImageSharp/Formats/Webp/WebpAnimationEncoder.cs:11]: Please write comments in English -- I'd like to be able to read them 😉 (w/o translator)

File: src/ImageSharp/Formats/Webp/WebpChunkParsingUtils.cs [hunk 2]
```diff
@@ -105,23 +104,16 @@ public static WebpImageInfo ReadVp8Header(MemoryAllocator memoryAllocator, Buffe
...
-            remaining,
-            memoryAllocator,
-            partitionLength)
-        {
-            Remaining = remaining
-        };
+        Vp8BitReader bitReader = new(stream, remaining, memoryAllocator, partitionLength) { Remaining = remaining };
 
-        return new WebpImageInfo()
+        return new()
```
Comment (gfoidl) [inline src/ImageSharp/Formats/Webp/WebpChunkParsingUtils.cs:116]: Nit: please specify the type here.

File: src/ImageSharp/Formats/Webp/WebpAnimationEncoder.cs (hunk 1 above)
Comment (Poker-sang) [reply src/ImageSharp/Formats/Webp/WebpAnimationEncoder.cs:11]: I'm sorry! I was going to remove this class, but forgot.

File: src/ImageSharp/Formats/Webp/WebpChunkParsingUtils.cs (hunk 2 above)
Comment (Poker-sang) [reply src/ImageSharp/Formats/Webp/WebpChunkParsingUtils.cs:116]: Ok.

**Comment by Poker-sang** [comment]
I'm sorry I used Resharper to refactor to `Specify created type`. Forgetting it leads to `SA1117`.

**Comment by JimBobSquarePants** [comment]
We need to throw some test images at this now. 

I would like to see some testing involving alpha blending as I think we've previously implemented it incorrectly. If it works anything like png the blend should take place per scanline where a temp buffer is used to fold the initial decode before blending it with the frame row. If you look at the latest changes to #2511 you can see how it is implemented there.

**Comment by Poker-sang** [comment]
Do you mean that I should modify the WebP decoder and encoder according to APNG? Especially to address issues that arise with Blending and Disposing. Then select some new test images?

**Comment by saucecontrol** [comment]
There's a good sample for alpha blending tests at https://bugs.chromium.org/p/chromium/issues/detail?id=364679 (landscape.webp linked at the top)

**Comment by JimBobSquarePants** [comment]
> Do you mean that I should modify the WebP decoder and encoder according to APNG? Especially to address issues that arise with Blending and Disposing. Then select some new test images?

Yes. 

Blending needs to work per scanline within the frame region against the current frame. 

We always build a frame using the previous frame data. We might then clear that data before populating the frame with the decoded pixel data. If that newly decoded data represents transparent pixels and you blend with the previous frame you then lose that transparency. What we should do is write the scanline to a temporary buffer and blend that buffer with the current frame. One of the tests png images alerted me to that issue. 

I've also noticed we're not capturing the disposal or blending data in the webp frame metadata, we need that for encoding.

**Comment by Poker-sang** [comment]
@JimBobSquarePants I tried to modify blend operation according to <https://github.com/SixLabors/ImageSharp/pull/2511/commits/66f444d200985720a9ff3e27a7a52d4722836b43>. But the decoder seems to have a problem whereas it didn't before. I'm not sure if I misunderstood what you meant before. Sorry if so.

Before:
image

After:
image

**Comment by JimBobSquarePants** [comment]
@Poker-sang I'll try to have a look tonight.

**Comment by JimBobSquarePants** [comment]
@Poker-sang Where is that test image from? I can't see it in the input folder

**Comment by JimBobSquarePants** [comment]
OK. Alpha blending works as expected using the provided test image using the correct approach.

**Review by JimBobSquarePants — APPROVED** [review]
@Poker-sang This looks great now. Supper happy! Thanks ever so much for your help here and elsewhere.

**Comment by Poker-sang** [comment]
@JimBobSquarePants Wow it's merged. Thank you for your help and quick response!

**Comment by Poker-sang** [comment]
Maybe we can close #1802 as well.

File: src/ImageSharp/Formats/Webp/WebpAnimationDecoder.cs [hunk 3]
```diff
@@ -291,42 +288,43 @@ private Buffer2D<TPixel> DecodeImageData<TPixel>(WebpFrameData frameData, WebpIm
...
+            // background color within the restore area.
+            PixelBlender<TPixel> blender =
+                PixelOperations<TPixel>.Instance.GetPixelBlender(PixelColorBlendingMode.Normal, PixelAlphaCompositionMode.SrcOver);
+
+            for (int y = 0; y < restoreArea.Height; y++)
+            {
+                Span<TPixel> framePixelRow = imageFramePixels.DangerousGetRowSpan(y);
+                Span<TPixel> decodedPixelRow = decodedImageFrame.DangerousGetRowSpan(y)[..restoreArea.Width];
+
+                blender.Blend<TPixel>(imageFrame.Configuration, framePixelRow, framePixelRow, decodedPixelRow, 1f);
```
Comment (Poker-sang) [inline src/ImageSharp/Formats/Webp/WebpAnimationDecoder.cs:317]: I used to blend pixels from the `previousFrame` and the `decodedImageFrame` here. No wonder it's a problem.

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Webp/WebpAnimationDecoder.cs:317]: Yeah, hopefully my comment in the code makes it clear why it has to work this way.

**Pull request merged**